Don't merge before release: 3.13 blog post (#260)
ember-learn/ember-blog: source/2019-09-20-ember-3-13-released.md

---
title: Ember 3.13 Released
author: Kenneth Larsen

> tags: Releases, 2019, 3, 3.13

Reviewer: Needs `Version 3.x` in this list so it shows up under the 3.x link on the blog!

```suggestion
tags: Releases, 2019, 3, 3.13, Version 3.x
```

Author: omg thanks!

Reviewer: Added this as a suggestion so you can trivially auto-apply it!

responsive: true
---

Today the Ember project is releasing version 3.13 of Ember.js, Ember Data, and Ember CLI. 

Additionally, versions 3.12 of Ember and Ember Data are now promoted to LTS, which stands for Long Term Support. An LTS version of Ember continues to receive security updates for 9 release cycles (54 weeks) and bugfixes for 6 cycles (36 weeks). LTS releases typically occur every four minor versions. The previous LTS versions for Ember and Ember Data were 3.4.

> If you are upgrading from the previous LTS version to 3.12 then you are getting access to these features:

Reviewer 2: When I first read this, I thought it was saying 3.6, 3.8, 3.10, 3.11 are all LTS versions, which confused me.

Reviewer 2: (I think the list is good/interesting, but the way it is worded/presented was a bit confusing to me)

* The `{{fn}}` and `{{on}}` modifiers

This release kicks off the 3.14 beta cycle for all sub-projects. We encourage our community (especially addon authors) to help test these beta builds and report any bugs before they are published as a final release in six weeks' time. The [ember-try](https://github.com/ember-cli/ember-try) addon is a great way to continuously test your projects against the latest Ember releases.

You can read more about our general release process here:

- [Release Dashboard](http://emberjs.com/builds/)
- [The Ember Release Cycle](http://emberjs.com/blog/2013/09/06/new-ember-release-process.html)
- [The Ember Project](http://emberjs.com/blog/2015/06/16/ember-project-at-2-0.html)
- [Ember LTS Releases](http://emberjs.com/blog/2016/02/25/announcing-embers-first-lts.html)

---

## Ember.js

Ember.js is the core framework for building ambitious web applications.

### Changes in Ember.js 3.13

Ember.js 3.13 is an incremental, backwards compatible release of Ember with bugfixes, performance improvements, and minor deprecations. There are five (5) new features, one (1) deprecation, and several bugfixes in this version.

#### New Features (5)

> **Tracked Properties and Tracked Property Updates (1 of 5)**

Reviewer: Suggestion for semantics: these are being **strongly emphasized** but what you really mean is "This is a section." These should be `#####` for `h5` tags. We need to add styles for that to look right, which I can certainly do if someone points me to where (and I can go back and update all previous blog posts using this format as well).

Author: You're right. It's been there because we don't have the styles for it, I believe. There is a redesign going on right now which I hope will resolve the issue

Tracked properties introduce a simpler and more ergonomic system for tracking state change in Ember applications. By taking advantage of new JavaScript features, tracked properties allow Ember to reduce its API surface area while producing code that is both more intuitive and less error-prone.

> During the Ember Octane preview period we encountered some issues with the current design for Tracked Properties that was proposed and accepted in [RFC 410](https://github.com/emberjs/rfcs/blob/master/text/0410-tracked-properties.md). The primary issues were specifically around interop between tracked properties, computed properties, and autotracking, with a few extra issues and inconsistencies surrounding these.

Reviewer: I believe there were also, and equally seriously, issues around interop with sync observers—and the fixes made with them have performance implications that people should know about here. cc. @pzuraq.

Reviewer 3: Yes, observer heavy applications should definitely keep an eye out for performance regressions with this release, and consider switching on the async-observers optional feature in order to mitigate if they find issues. We should note that here for sure.

Reviewer 2: Should we just add async observers as a separate top-level items?

With Ember.js 3.13 you now use this improved version of tracked properties. Have a look at the [guides] or [RFC 410](https://github.com/emberjs/rfcs/blob/master/text/0410-tracked-properties.md) to learn more about tracked properties. You can also have a look at [RFC 478](https://github.com/emberjs/rfcs/blob/master/text/0478-tracked-properties-updates.md) to see what the tracked properties update is about.

**Component Templates Co-location (2 of 5)**

With component templates co-location it is now possible to place a component's JavaScript file and it's template in the same directory on the file system.

> [The RFC](https://github.com/emberjs/rfcs/blob/master/text/0481-component-templates-co-location.md) contains a detailed explanaition for this feature.

Reviewer 2: Only the low-level APIs: `setComponentTemplate()`, `getComponentTemplate()` and `templateOnlyComponent()` are enabled in this release. The supporting code for the actual co-location feature is included in the release, but is only enabled with octane apps... this feature will be enabled for all apps in the next version.

Reviewer: @chancancode @kennethlarsen see my recommendation in the overall review comment I'm leaving; I'd suggest we note that users can try it out by testing Octane on the beta release channel.

**Add `updateHook` component-manager capability (3 of 5)**

There is a new capability for component managers which allows you to opt-in or opt-out of whether to have an `updateComponent` hook on your component manager.

If you're not an addon author or maintain a component manager then chances are that you will not need use this new feature.

> **Component-class generator (4 of 5)**

Reviewer 2: ```suggestion
**`component-class` generator (4 of 5)**
```

> When generating a component using `ember generate component modal` it will now by default use the `--no-component-class` flag. This means that when you generate a component it will only generate the template file and the test file.

Reviewer 2: This is only true in Octane apps

> If you want a component class file when generating the component you can use the `--with-component-class` flag.

Reviewer 2: ```suggestion
In the cases where you need a component class file, you can either pass the `--with-component-class` flag to the `component` generator, or add one after the fact with the new `ember generate component-class modal` generator.
```

In Ember 3.14, Octane will be the recommended way to use Ember and new apps will have Octane's optional features enabled by default. The [guides](https://octane-guides-preview.emberjs.com/release/) and [tutorial](https://octane-guides-preview.emberjs.com/release/tutorial/00-part-1/) will show Octane examples, and codemods will be available to help users migrate to Octane.

---

You can try out the Octane preview by disabling legacy behavior and specifying the Octane edition in `.ember-cli.js`.

```bash
ember feature:disable jquery-integration
ember feature:enable template-only-glimmer-components
ember feature:disable application-template-wrapper
```

If you need more information on how to migrate away from these legacy features, check out the [Octane release plan](https://blog.emberjs.com/2019/08/15/octane-release-plan.html) blog post.

To opt in to the Octane preview, add the following lines to the beginning of `.ember-cli.js`.

```js
// .ember-cli.js
const { setEdition } = require('@ember/edition-utils');

> setEdition('octane');

Reviewer 4: Precisely what does this do?

Reviewer 5: I think this switches the blueprints. If that's true, we should specifically say that.

### Changes in Ember CLI 3.13

No new features in Ember CLI 3.13.

---

For more details on the changes in Ember CLI 3.13 and detailed upgrade
instructions, please review the [Ember CLI  3.13.0 release page](https://github.com/ember-cli/ember-cli/releases/tag/v3.13.0).

## Ember 3.12 is an LTS Release

If you are upgrading from the previous LTS version (3.8) to 3.12 then you are getting access to these features:

* `router` service 
* Support for native classes
* The `array` helper
* Element Modifier Manager (and modifiers in general)
* Nested angle bracket component invocation
* Angle bracket invocation for built-in components
* Support for native decorators
* The `{{fn}}` and `{{on}}` modifiers

## 3.14 Beta

This release kicks off the 3.14 beta cycle for all sub-projects. We encourage our community (especially addon authors) to help test these beta builds and report any bugs before they are published as a final release in six weeks' time. The [ember-try](https://github.com/ember-cli/ember-try) addon is a great way to continuously test your projects against the latest Ember releases.

> Among other things, the [`@model` named argument](https://github.com/emberjs/rfcs/blob/master/text/0523-model-argument-for-route-templates.md) will be available to route templates. While not officially an "Octane feature", this feature will available for the first time in the 3.14 release, which coincides with the Octane release. Since the Octane guides and documentation are targeting the same version, you will see them using this feature, but please not that the feature is not available in the 3.13 Octane preview.

Reviewer 6: ```suggestion
Among other things, the [`@model` named argument](https://github.com/emberjs/rfcs/blob/master/text/0523-model-argument-for-route-templates.md) will be available to route templates. While not officially an "Octane feature", this feature will available for the first time in the 3.14 release, which coincides with the Octane release. Since the Octane guides and documentation are targeting the same version, you will see them using this feature, but please note that the feature is not available in the 3.13 Octane preview.
```

---
title: Ember 3.13 Released (Octane Preview)
author: Kenneth Larsen, Tom Dale, Yehuda Katz
tags: Releases, 2019, 3, 3.13, Version 3.x
responsive: true
---

Today, the Ember project is releasing version 3.13 of Ember.js, Ember Data, and Ember CLI.

> Ember 3.13 is the final release of Ember before Octane. Ember 3.13 is also the Octane preview. The preview allows adventurous community members to help us put the final round of polish on the edition before Octane becomes official in 3.14.

Reviewer 5: ```suggestion
Ember 3.13 is the final release of Ember before Octane. Ember 3.13 is also the Octane preview. The preview allows adventurous community members to start using its stable features and help put the final round of polish on supporting tools and documentation before Octane becomes official in 3.14. You can learn more about what to expect from Octane and how to prepare your apps in a [previous blog post](https://blog.emberjs.com/2019/08/15/octane-release-plan.html).
```

Additionally, version 3.12 of Ember is now promoted to LTS (Long Term Support). An LTS version of Ember continues to receive security updates for 9 release cycles (54 weeks) and bugfixes for 6 cycles (36 weeks). LTS releases typically occur every four minor versions. The previous LTS version of Ember was 3.8.

## The Octane Preview

The Octane features in 3.13 are stable public APIs with our normal SemVer guarantees. If you build an app using APIs in the Octane preview, it will continue to work in future versions.

So why is this a preview and not the final Octane release?

> While these APIs are stable, not all Octane features are compatible with popular addons yet. We are also still updating documentation, and important tools like the Ember Inspector do not yet work with some Octane features.

Reviewer 5: ```suggestion
A big part of Octane is about delivering an excellent developer experience, whether you are a beginner or a power user. While these APIs are stable, not all Octane features are compatible with popular addons yet. We are also still updating documentation, and important tools like the Ember Inspector do not yet work with some Octane features. Octane is not complete until these supporting pieces are in place.
```

> In Ember 3.14, Octane will be the recommended way to use Ember and new apps will have Octane's optional features enabled by default. The [guides](https://octane-guides-preview.emberjs.com/release/) and [tutorial](https://octane-guides-preview.emberjs.com/release/tutorial/00-part-1/) will show Octane examples, and codemods will be available to help users migrate to Octane.

Reviewer 5: ```suggestion
Then, when Ember 3.14 arrives, Octane will be the recommended way to use Ember and new apps will have Octane's optional features enabled by default. The guides and tutorials will show Octane examples, and codemods will be available to help users migrate to Octane. Feel free to take a look at the [current draft](https://emberjs.com/editions/octane) of these learning resources, which are deployed to a temporary URL.
```

> The Ember tutorial has already been completely rewritten for Octane, and the easiest way to understand the big picture of Octane, especially if you're already an experienced Ember developer, is to [work through the new tutorial](https://octane-guides-preview.emberjs.com/release/tutorial/00-part-1/).

Reviewer 5: ```suggestion
The Ember tutorial has already been completely rewritten for Octane, and the easiest way to understand the big picture of Octane, especially if you're already an experienced Ember developer, is to [work through the new tutorial](https://emberjs.com/editions/octane).
```

We should not link to the preview URL, since it will be removed when Octane ships.

ember feature:enable template-only-glimmer-components
ember feature:disable application-template-wrapper
```

If you need more information on how to migrate away from these legacy features, check out the [Octane release plan](https://blog.emberjs.com/2019/08/15/octane-release-plan.html) blog post.

To opt in to the Octane preview, add the following lines to the beginning of `.ember-cli.js`.

```js
// .ember-cli.js
const { setEdition } = require('@ember/edition-utils');

setEdition('octane');
```

---

## Ember.js

Ember.js is the core framework for building ambitious web applications.

### Changes in Ember.js 3.13

Ember.js 3.13 is an incremental, backwards compatible release of Ember with bugfixes, performance improvements, and minor deprecations. There are five (5) new features, one (1) deprecation, and several bugfixes in this version.

#### New Features (5)

**Tracked Properties and Tracked Property Updates (1 of 5)**

Tracked properties introduce a simpler and more ergonomic system for tracking state change in Ember applications. Tracked properties allow Ember to reduce its API surface area while producing code that is both more intuitive and less error-prone.

> Have a look at the [guides] to learn more about tracked properties. For design details, check out [RFC 410](https://github.com/emberjs/rfcs/blob/master/text/0410-tracked-properties.md) and [RFC 478](https://github.com/emberjs/rfcs/blob/master/text/0478-tracked-properties-updates.md).

Reviewer 5: ```suggestion
Have a look at the [guides] to learn more about tracked properties. For design details, check out [RFC 410](https://github.com/emberjs/rfcs/blob/master/text/0410-tracked-properties.md) and [RFC 478](https://github.com/emberjs/rfcs/blob/master/text/0478-tracked-properties-updates.md).
```
This should link to the API docs for tracked, ideally. If we have to, we can link to https://emberjs.com/editions/octane and say "have a look at the draft guides...."

Ember CLI is the command line interface for managing and packaging Ember.js applications.

### Upgrading Ember CLI

You may upgrade Ember CLI easily using the ember-cli-update project:

```bash
npm install -g ember-cli-update
ember-cli-update
```

This utility will help you to update your app or add-on to the latest Ember CLI version. You will probably encounter merge conflicts, in which the default behavior is to let you resolve conflicts on your own. For more information on the `ember-cli-update` project, see [the github README](https://github.com/ember-cli/ember-cli-update).

While it is recommended to keep Ember CLI versions in sync with Ember and Ember Data, this is not required. After updating ember-cli, you can keep your current version(s) of Ember or Ember Data by editing `package.json` to revert the changes to the lines containing `ember-source` and `ember-data`.

### Changes in Ember CLI 3.13

No new features in Ember CLI 3.13.

---

For more details on the changes in Ember CLI 3.13 and detailed upgrade
instructions, please review the [Ember CLI  3.13.0 release page](https://github.com/ember-cli/ember-cli/releases/tag/v3.13.0).

## Ember 3.12 is an LTS Release

> If you are upgrading from the previous LTS version (3.8) to 3.12 then you are getting access to these features:

Reviewer 5: love this section, thank you!

> * `router` service 

Reviewer 4: The router service has been included since Ember 2.12

Suggest saying specifically how it was improved.

> * Support for native classes

Reviewer 4: ES classes have been supported since August 2017, when [RFC 240](https://github.com/emberjs/rfcs/pull/240) was accepted.

Suggest saying [Native Class Constructor Update RFC](https://github.com/emberjs/rfcs/blob/master/text/0337-native-class-constructor-update.md) implemented

Reviewer 3: Native classes were not officially supported until v3.6 (RFC implementation was incomplete + redesign occurred, no blog post announcement, no API docs). 3.10 adds decorators, which made native classes viable/recommendable for most users, so maybe it should say "Support for native classes + decorators"

* The `array` helper
* Element Modifier Manager (and modifiers in general)
* Nested angle bracket component invocation
* Angle bracket invocation for built-in components

> * Support for native decorators

Reviewer 4: I don't think any browser has native decorators. Reword?
